rfd/0030-upgrade-system.md

```diff
@@ -0,0 +1,444 @@
+teleport ecosystem. With this in mind, the upgrade system will be designed with
+the aspirational goal of being resilient to compromise of any single machine,
+secret, account, etc (in this context, "resilient" just means "won't result in
+successful execution of a malicious installation").  In practice, this level of
+fault-tolerance is essentially impossible to achieve, but compartmentalization
+and redundancy will be primary guiding principals.
+
+- The upgrade system must be *reliable*.  Specifically, the upgrade system must
+be resilient to intermittent failures and self-healing to the greatest extent possible.
+With this in mind, the upgrade system will embrace a kubernetes-esque model of enacting
+change; components will continually attempt to reconcile the current state of the system
+with some desired final state.  Care will also be taken to maximize cross-version and
+cross-implementation compatibility.
+
+- The upgrade system should be *modular* and *extensible*.  It should be easy to extend
+the behavior of the upgrade system both as a user and as a maintainer.  The individual
+components of the upgrade system must have simple and well-defined responsibilities.
+It should be easy to write new components, and to hook into external systems.  Components
+should also be highly opinionated about their domain of responsibility, while maintaining
+as little opinion as possible about other domains (this extends to mundane things, e.g. we
+will prefer purpose-specific validation and selection of defaults for resources, rather than
+a monolithic `CheckAndSetDefaults` operation as is common in much of teleport's internals).
+
+### Attack Scenarios
+
+#### Malicious Distribution
+
+**Scenario**: An attacker successfully leverages the upgrade system to directly distribute
+malicious non-teleport software (or a malicious fork of teleport), causing existing teleport
+clusters to install said malicious software on one or more servers (aka: the doomsday scenario).
+
+We're going to roughly divide the possible avenues of attack into two families, direct and
+indirect compromise.  In the case of direct compromise, the attacker successfully causes
+"legitimate" distribution infrastructure to perform functions that result in the installation
+of malicious software.  This may be via compromise of supply chain, software/libraries,
+accounts, or even individuals. In the case of indirect compromise, the attacker successfully
+tricks clients into believing that they are interacting with, or received a package from,
+some trusted source, when that is not the case.  This may occur due to the compromise of a
+cryptographic key or CA, manipulation of a client into using insecure protocols, redirection
+of a client to an illigitimate download point, injection of malicious public key into a root
+of trust, etc.
+
+Direct compromise will be mitigated via compartmentalization.  In particular, we will divide
```

Comment:
What would the real-world usage of the split system be like?

Let's assume that the CI/CD system produces the artifacts and publishes to S3. We have to have an independent release and build step when someone or something publishes the signatures to the independent registry? I think that the devil is in the details here.

We could compute the signature independently of the CI/CD system if we use [deterministic builds](https://github.com/golang/go/issues/32782) from golang.

Reply:
MVP: We manually update static index configuration files after artifacts are successfully published (maybe we compute checksums as part of build pipeline, but otherwise all manual).  Totally fine if index trails behind available releases, so this step is async.

Long term: definitely deterministic builds w/ fully isolated build pipeline.

Reply:
@fspmarshall what prevents us from doing fully deterministic builds right from the start if Go supports it? Manual steps are quite error prone, wondering if we will save more time by automating it right from the start vs manual updating? E.g. if I send a PR to update the index, someone has also to validate it. We can may be even host index on github repo :D

Side thought: if index is as easy, we can sign the index and host in on github repo itself

Reply:
Hosting on github will let people have their own mirrors trivially, by using git clone as well :)

Reply:
Couple of other random thoughts:

* May be time to experiment with formats other than YAML. We started using TOML with plugins, pretty happy with it's simplicity and lack of dealing with offsets :)
* Another random thought is to make index self-evaluating. E.g it could be a lisp program that you can generate and run, and it will output it's own verification status :) But that may be too exotic for non-lisp users :D

Reply:
> what prevents us from doing fully deterministic builds right from the start

Not sure.  I'm operating under the assumption that reproducible builds might be more of an undertaking than is practical for the MVP timeline.  I'll take some time to try to verify this assumption tho.  Maybe they aren't as hard as I think.  

> we can sign the index and host in on github repo itself

I'd been planning to have some kind of API for getting specific version ranges, but being able to mirror by forking is definitely appealing.

If we're going to pull down the index from github, I think it would make the most sense if, at least by default, clients pull down a file that is a "sliding window" of the currently supported releases.  Append-only log is nice (and we could totally have one of those in the repo too), but it would be weird to be pulling down manifests for releases that are years out of date on a daily basis.

> May be time to experiment with formats other than YAML. We started using TOML with plugins, pretty happy with it's simplicity and lack of dealing with offsets :)

Love TOML, so I'm on board with this in theory.  Worth noting, TOML is specifically designed to represent very flat data, so we'd probably need to be careful to avoid nesting (e.g. move the `sums` mapping into the outer target spec, stuff like that).  We could also break up the log into many small files, but that would definitely increase complexity if we want to be pulling the log as individual files from the github REST api.  Hypothetical syntax of a monolithic `releases.toml`:

```toml
[[release]]
version = '1.2.3'
stable = true
security-patch = false

[[release.target]]
flavor = 'oss'
os = 'linux'
arch = 'amd64'
tgz-sha3-256 = '...'
rpm-sha3-256 = '...'
deb-sha3-256 = '...'

[[release.target]]
flavor = 'ent'
os = 'linux'
arch = 'amd64'
tgz-sha3-256 = '...'
rpm-sha3-256 = '...'
deb-sha3-256 = '...'

[[release]]
version = '2.3.4'
stable = true
security-patch = true

[[release.target]]
flavor = 'oss'
os = 'linux'
arch = 'amd64'
tgz-sha3-256 = '...'
rpm-sha3-256 = '...'
deb-sha3-256 = '...'

[[release.target]]
flavor = 'ent'
os = 'linux'
arch = 'amd64'
tgz-sha3-256 = '...'
rpm-sha3-256 = '...'
deb-sha3-256 = '...'
```


> make index self-evaluating. E.g it could be a lisp program that you can generate and run, and it will output it's own verification status :) But that may be too exotic for non-lisp users :D

Just add your lisp program as a tag to the release manifest and write a custom upgrade controller :wink:

Reply:
may be then instead of one static index file have a folder with releases and signatures. have the root index file that signs the sub indexes to keep track of the integrity

Reply:
I feel like we reinvent [TUF](https://theupdateframework.io/) here :)

Reply:
@fspmarshall have you considered the update framework's format may be?